framework/src/Volo.Abp.AspNetCore/Volo/Abp/AspNetCore/Tracing/AbpCorrelationIdMiddleware.cs

```diff
@@ -20,16 +21,37 @@ public class AbpCorrelationIdMiddleware : IMiddleware, ITransientDependency
 
     public async Task InvokeAsync(HttpContext context, RequestDelegate next)
     {
-        var correlationId = _correlationIdProvider.Get();
+        var correlationId = GetCorrelationIdFromRequest(context);
 
-        try
+        using (_correlationIdProvider.Change(correlationId))
         {
-            await next(context);
+            try
+            {
+                await next(context);
+            }
+            finally
+            {
+                CheckAndSetCorrelationIdOnResponse(context, _options, correlationId);
+            }
         }
-        finally
+    }
+
+    protected virtual string GetCorrelationIdFromRequest(HttpContext context)
+    {
+        string correlationId = context.Request.Headers[_options.HttpHeaderName];
+        if (correlationId.IsNullOrEmpty())
         {
-            CheckAndSetCorrelationIdOnResponse(context, _options, correlationId);
+            lock (context.Request.Headers)
+            {
```

Comment:
Shouldn't we try to get it again before the next `if`, like `correlationId = context.Request.Headers[_options.HttpHeaderName];`?

Reply:
However, I think `lock` is completely unnecessary. Because, an HTTP request is not multi-thread and there is no race condition.